[request] Add support for SEV provisioning in Azure

**Please describe problem to be solved**  
Currently Skyplane doesn't support provision confidential VMs in Azure.

```
⠏ Provisioning VMs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/1 0:00:2921:05:49 [ERROR] Error running _provision_task: (BadRequest) The VM size 'Standard_DC2as_v5' is not supported for creation of VMs and Virtual
Machine Scale Set with '<NULL>' security type.
Code: BadRequest
Message: The VM size 'Standard_DC2as_v5' is not supported for creation of VMs and Virtual Machine Scale Set with '<NULL>' security type.
Traceback (most recent call last):
  File "/home/zzz/workspace/skyplane/skyplane/cli/cli_transfer.py", line 353, in cp
    dp.provision(spinner=True)
  File "/home/zzz/workspace/skyplane/skyplane/api/dataplane.py", line 126, in provision
    uuids = self.provisioner.provision(
  File "/home/zzz/workspace/skyplane/skyplane/api/provisioner.py", line 205, in provision
    results: List[Tuple[ProvisionerTask, compute.Server]] = do_parallel(
  File "/home/zzz/workspace/skyplane/skyplane/utils/fn.py", line 57, in do_parallel
    args, result = future.result()
  File "/home/zzz/.pyenv/versions/3.8.10/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()
  File "/home/zzz/.pyenv/versions/3.8.10/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/home/zzz/.pyenv/versions/3.8.10/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/zzz/workspace/skyplane/skyplane/utils/fn.py", line 43, in wrapped_fn
    return args, func(args)
  File "/home/zzz/workspace/skyplane/skyplane/api/provisioner.py", line 154, in _provision_task
    server = self.azure.provision_instance(task.region, task.vm_type, use_spot_instances=task.spot, tags=task.tags)
  File "/home/zzz/workspace/skyplane/skyplane/utils/imports.py", line 33, in wrapped
    return fn(*modules_imported, *args, **kwargs)
  File "/home/zzz/workspace/skyplane/skyplane/compute/azure/azure_cloud_provider.py", line 342, in provision_instance
    poller = compute_client.virtual_machines.begin_create_or_update(
  File "/home/zzz/.pyenv/versions/3.8.10/envs/skyplane/lib/python3.8/site-packages/azure/core/tracing/decorator.py", line 78, in
wrapper_use_tracer
    return func(*args, **kwargs)
  File
"/home/zzz/.pyenv/versions/3.8.10/envs/skyplane/lib/python3.8/site-packages/azure/mgmt/compute/v2022_11_01/operations/_virtual_machines_operation
s.py", line 1358, in begin_create_or_update
    raw_result = self._create_or_update_initial(
  File
"/home/zzz/.pyenv/versions/3.8.10/envs/skyplane/lib/python3.8/site-packages/azure/mgmt/compute/v2022_11_01/operations/_virtual_machines_operation
s.py", line 1229, in _create_or_update_initial
    raise HttpResponseError(response=response, error_format=ARMErrorFormat)
azure.core.exceptions.HttpResponseError: (BadRequest) The VM size 'Standard_DC2as_v5' is not supported for creation of VMs and Virtual Machine
Scale Set with '<NULL>' security type.
Code: BadRequest
Message: The VM size 'Standard_DC2as_v5' is not supported for creation of VMs and Virtual Machine Scale Set with '<NULL>' security type.

```

**(Optional): Suggest a solution**  
Rework on the provisioning code to support the creation of CVM
* Details of the technical implementation
If the vm instance is CVM, we need to set the security profile and the os profile to a proper value.
* Tradeoffs made in design decisions
* Caveats and considerations for the future)
TODO add CVM provisioning for other cloud providers as well.

--- a/skyplane/compute/azure/azure_cloud_provider.py
+++ b/skyplane/compute/azure/azure_cloud_provider.py
@@ -97,7 +97,7 @@ class AzureCloudProvider(CloudProvider):
         return "unknown"
 
     @staticmethod
-    def use_confidential_instances(instance_name: str) -> bool:
+    def is_confidential_instance_type(instance_name: str) -> bool:
         # Lists of confidential vms can be found here:
         # https://learn.microsoft.com/en-us/azure/confidential-computing/virtual-machine-solutions-amd#sizes
         # For example, `Standard_DC4ads_v5`, `Standard_EC64ads_v5` are confidential instances.
@@ -395,7 +395,7 @@ class AzureCloudProvider(CloudProvider):
                         # use spot instances if use_spot_instances is set
                         "priority": "Spot" if use_spot_instances else "Regular",
                     }
-                    if self.use_confidential_instances(vm_size):
+                    if self.is_confidential_instance_type(vm_size):
                         vm_params["security_profile"] = {
                             "security_type": "ConfidentialVM",
                             "uefi_settings": {"v_tpm_enabled": True, "secure_boot_enabled": True},

--- a/skyplane/compute/azure/azure_cloud_provider.py
+++ b/skyplane/compute/azure/azure_cloud_provider.py
@@ -97,6 +97,17 @@ class AzureCloudProvider(CloudProvider):
         return "unknown"
 
     @staticmethod
+    def use_confidential_instances(instance_name: str) -> bool:
+        # Lists of confidential vms can be found here:
+        # https://learn.microsoft.com/en-us/azure/confidential-computing/virtual-machine-solutions-amd#sizes
+        # For example, `Standard_DC4ads_v5`, `Standard_EC64ads_v5` are confidential instances.
+        sz = instance_name.split("_")[1]
+        # remove the digits from the string
+        sz = re.sub(r"\d", "", sz)
+        if sz.upper() in {"DCAS", "DCADS", "ECAS", "ECADS"}:
+            return True
+
+    @staticmethod
     def lookup_valid_instance(region: str, instance_name: str) -> Optional[str]:
         sku_mapping = AzureAuthentication.get_sku_mapping()
         if instance_name in sku_mapping[region]:
@@ -339,54 +350,70 @@ class AzureCloudProvider(CloudProvider):
         with Timer("Creating Azure VM"):
             with self.provisioning_semaphore:
                 try:
+                    vm_params = {
+                        "location": location,
+                        "tags": tags,
+                        "hardware_profile": {"vm_size": self.lookup_valid_instance(location, vm_size)},
+                        "storage_profile": {
+                            # "image_reference": {
+                            #     "publisher": "canonical",
+                            #     "offer": "0001-com-ubuntu-server-focal",
+                            #     "sku": "20_04-lts",
+                            #     "version": "latest",
+                            # },
+                            "image_reference": {
+                                "publisher": "microsoft-aks",
+                                "offer": "aks",
+                                "sku": "aks-engine-ubuntu-1804-202112",
+                                "version": "latest",
+                            },
+                            "os_disk": {"create_option": "FromImage", "delete_option": "Delete"},
+                        },
+                        "os_profile": {
+                            "computer_name": AzureServer.vm_name(name),
+                            "admin_username": uname,
+                            "linux_configuration": {
+                                "disable_password_authentication": True,
+                                "ssh": {"public_keys": [{"path": f"/home/{uname}/.ssh/authorized_keys", "key_data": pub_key}]},
+                            },
+                        },
+                        "network_profile": {"network_interfaces": [{"id": nic_result.id}]},
+                        # give VM managed identity w/ user assigned identity
+                        "identity": {
+                            "type": ResourceIdentityType.user_assigned,
+                            "user_assigned_identities": [
+                                {
+                                    # code from: https://github.com/ray-project/ray/pull/7080/files#diff-0f1bb1da82d112b850a85c1b7b1876e50efd5a7400c62a5c4de334f494d3bf46R222-R233
+                                    "key": f"/subscriptions/{cloud_config.azure_subscription_id}/resourceGroups/skyplane/providers/Microsoft.ManagedIdentity/userAssignedIdentities/skyplane_umi",
+                                    "value": {
+                                        "principal_id": cloud_config.azure_principal_id,
+                                        "client_id": cloud_config.azure_client_id,
+                                    },
+                                }
+                            ],
+                        },
+                        # use spot instances if use_spot_instances is set
+                        "priority": "Spot" if use_spot_instances else "Regular",
+                    }
+                    if self.use_confidential_instances(vm_size):
+                        vm_params["security_profile"] = {"security_type": "ConfidentialVM", "uefi_settings": {"v_tpm_enabled": True}}
+                        vm_params["storage_profile"]["os_disk"].update(
+                            {
+                                "managed_disk": {
+                                    "security_profile": {"security_encryption_type": "VMGuestStateOnly"},
+                                }
+                            }
+                        )
+                        vm_params["storage_profile"]["image_reference"] = {
+                            "publisher": "Canonical",
+                            "offer": "0001-com-ubuntu-confidential-vm-focal",
+                            "sku": "20_04-lts-cvm",
+                            "version": "latest",
+                        }
                     poller = compute_client.virtual_machines.begin_create_or_update(
                         resource_group,
                         AzureServer.vm_name(name),
-                        {
-                            "location": location,
-                            "tags": tags,
-                            "hardware_profile": {"vm_size": self.lookup_valid_instance(location, vm_size)},
-                            "storage_profile": {
-                                # "image_reference": {
-                                #     "publisher": "canonical",
-                                #     "offer": "0001-com-ubuntu-server-focal",
-                                #     "sku": "20_04-lts",
-                                #     "version": "latest",
-                                # },
-                                "image_reference": {
-                                    "publisher": "microsoft-aks",
-                                    "offer": "aks",
-                                    "sku": "aks-engine-ubuntu-1804-202112",
-                                    "version": "latest",
-                                },
-                                "os_disk": {"create_option": "FromImage", "delete_option": "Delete"},
-                            },
-                            "os_profile": {
-                                "computer_name": AzureServer.vm_name(name),
-                                "admin_username": uname,
-                                "linux_configuration": {
-                                    "disable_password_authentication": True,
-                                    "ssh": {"public_keys": [{"path": f"/home/{uname}/.ssh/authorized_keys", "key_data": pub_key}]},
-                                },
-                            },
-                            "network_profile": {"network_interfaces": [{"id": nic_result.id}]},
-                            # give VM managed identity w/ user assigned identity
-                            "identity": {
-                                "type": ResourceIdentityType.user_assigned,
-                                "user_assigned_identities": [
-                                    {
-                                        # code from: https://github.com/ray-project/ray/pull/7080/files#diff-0f1bb1da82d112b850a85c1b7b1876e50efd5a7400c62a5c4de334f494d3bf46R222-R233
-                                        "key": f"/subscriptions/{cloud_config.azure_subscription_id}/resourceGroups/skyplane/providers/Microsoft.ManagedIdentity/userAssignedIdentities/skyplane_umi",
-                                        "value": {
-                                            "principal_id": cloud_config.azure_principal_id,
-                                            "client_id": cloud_config.azure_client_id,
-                                        },
-                                    }
-                                ],
-                            },
-                            # use spot instances if use_spot_instances is set
-                            "priority": "Spot" if use_spot_instances else "Regular",
-                        },
+                        vm_params,
                     )
                     try:
                         vm_result = poller.result()

--- a/skyplane/compute/azure/azure_cloud_provider.py
+++ b/skyplane/compute/azure/azure_cloud_provider.py
@@ -396,11 +396,14 @@ class AzureCloudProvider(CloudProvider):
                         "priority": "Spot" if use_spot_instances else "Regular",
                     }
                     if self.use_confidential_instances(vm_size):
-                        vm_params["security_profile"] = {"security_type": "ConfidentialVM", "uefi_settings": {"v_tpm_enabled": True}}
+                        vm_params["security_profile"] = {
+                            "security_type": "ConfidentialVM",
+                            "uefi_settings": {"v_tpm_enabled": True, "secure_boot_enabled": True},
+                        }
                         vm_params["storage_profile"]["os_disk"].update(
                             {
                                 "managed_disk": {
-                                    "security_profile": {"security_encryption_type": "VMGuestStateOnly"},
+                                    "security_profile": {"security_encryption_type": "DiskWithVMGuestState"},
                                 }
                             }
                         )

